ta/pkcs11/src/processing.c

```diff
@@ -450,3 +450,186 @@ enum pkcs11_rc entry_processing_step(struct pkcs11_client *client,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = params;
+	TEE_Param *out = params + 2;
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	struct serialargs ctrlargs = { };
+	struct pkcs11_session *session = NULL;
+	struct pkcs11_attribute_head *proc_params = NULL;
+	struct pkcs11_object_head *template = NULL;
+	uint32_t parent_handle = 0;
+	uint32_t obj_handle = 0;
+	struct pkcs11_object *parent = NULL;
+	struct obj_attrs *head = NULL;
+	size_t template_size = 0;
+	enum processing_func function = PKCS11_FUNCTION_DERIVE;
+
+	if (!client || ptypes != exp_pt ||
+	    out->memref.size != sizeof(obj_handle))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rc = serialargs_get_session_from_handle(&ctrlargs, client, &session);
+	if (rc)
+		return rc;
+
+	rc = serialargs_get(&ctrlargs, &parent_handle, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	rc = serialargs_alloc_get_one_attribute(&ctrlargs, &proc_params);
+	if (rc)
+		goto out_free;
+
+	rc = serialargs_alloc_get_attributes(&ctrlargs, &template);
+	if (rc)
+		goto out_free;
+
+	if (serialargs_remaining_bytes(&ctrlargs)) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out_free;
+	}
+
+	/* Return error if processing already active */
+	rc = get_ready_session(session);
+	if (rc)
+		goto out_free;
+
+	/* Check parent handle */
+	parent = pkcs11_handle2object(parent_handle, session);
+	if (!parent) {
+		rc = PKCS11_CKR_KEY_HANDLE_INVALID;
+		goto out_free;
+	}
+
+	/* Check if mechanism supplied is supported for key derivation */
+	switch (proc_params->id) {
+	case PKCS11_CKM_AES_ECB_ENCRYPT_DATA:
+	case PKCS11_CKM_AES_CBC_ENCRYPT_DATA:
+		break;
+	default:
+		rc = PKCS11_CKR_MECHANISM_INVALID;
+		goto out;
+	}
+
+	/* Set the processing state to active */
+	rc = set_processing_state(session, function, parent, NULL);
+	if (rc)
+		goto out_free;
+
+	/* Check if mechanism can be used for derivation function */
+	rc = check_mechanism_against_processing(session, proc_params->id,
+						function,
+						PKCS11_FUNC_STEP_INIT);
+	if (rc)
+		goto out;
+
+	/*
+	 * Check if base/parent key has CKA_DERIVE set and its key type is
+	 * compatible with the mechanism passed
+	 */
+	rc = check_parent_attrs_against_processing(proc_params->id, function,
+						   parent->attributes);
+	if (rc)
+		goto out;
+
+	/* Check access of base/parent key */
+	rc = check_access_attrs_against_token(session, parent->attributes);
+	if (rc)
+		goto out;
+
+	template_size = sizeof(*template) + template->attrs_size;
+	/*
+	 * Prepare a clean initial state for the requested object attributes.
+	 * using base/parent key attributes.
+	 * Free temporary template once done.
+	 */
+	rc = create_attributes_from_template(&head, template, template_size,
+					     parent->attributes,
+					     PKCS11_FUNCTION_DERIVE,
+					     proc_params->id,
+					     PKCS11_CKO_UNDEFINED_ID);
+	if (rc)
+		goto out;
+
+	TEE_Free(template);
+	template = NULL;
+
```

Comment:
For consistency with other processing sequence used to generate keys, i think it would be nice to add an inline comment here stating `check_created_attrs()` is called later once key size is known.